### PR TITLE
Extend IJsDialogHandler To Include OnBeforeUnload Dialogs

### PR DESCRIPTION
**Pull request by daveaton**
Extended IJsDialogHandler to catch on OnBeforeUnload Javascript Dialog messages.

This should fix the issue with #712

### CODE REVIEW

**Comment by amaitland** [comment]
@daveaton It looks like your pretty close. I have a few questions, I'll comment inline.

File: CefSharp.Core/Internals/ClientAdapter.cpp [hunk 1]
```diff
@@ -446,6 +446,29 @@ namespace CefSharp
...
+            }
+
+            bool result;
+            bool handled = false;
+
+            String^ resultString = nullptr;
+            handled = handler->OnJSBeforeUnload(_browserControl, StringUtils::ToClr(message_text), is_reload, resultString);
+            if (handled)
+            {
+                callback->Continue(is_reload, StringUtils::ToNative(resultString));
```
Comment (amaitland) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: Should `is_reload` be passed as first param here?

File: CefSharp/IJsDialogHandler.cs [hunk 2]
```diff
@@ -9,5 +9,6 @@ public interface IJsDialogHandler
         bool OnJSAlert(IWebBrowser browser, string url, string message);
         bool OnJSConfirm(IWebBrowser browser, string url, string message, out bool retval);
         bool OnJSPrompt(IWebBrowser browser, string url, string message, string defaultValue, out bool retval, out string result);
+        bool OnJSBeforeUnload(IWebBrowser browser, string message, out bool is_reload, out string result);
```
Comment (amaitland) [inline CefSharp/IJsDialogHandler.cs]: Can you rename `is_reload` to `isReload` please :+1:

File: CefSharp.Core/Internals/ClientAdapter.cpp [hunk 3]
```diff
@@ -446,6 +446,29 @@ namespace CefSharp
...
+            }
+
+            return handled;
+
```
Comment (amaitland) [inline CefSharp.Core/Internals/ClientAdapter.cpp]: Please remove extra line. I'm being pedantic I know :smile:

**Comment by daveaton** [comment]
Hum...  just learning how Git Hub works, it seems that it automatically pulls in changes, I thought I had to update the pull request some how..  but if this is the case doing a 1 to 1 master is not a good idea.. I should have created a branch and let you guys apply to master here... or am I wrong ?  If I make changes later on will it not update this pull request automatically ?

**Comment by amaitland** [comment]
Any change you push to a branch is included into a PR up until the point of it being merged. Long as you don't need to make other changes on your `master` branch, then this shouldn't be a problem :+1:

File: CefSharp3.sln.DotSettings [hunk 4]
```diff
@@ -98,6 +98,7 @@
 	<s:Boolean x:Key="/Default/Environment/SearchAndNavigation/GotoSingleUsageImmediately/@EntryValue">False</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SearchAndNavigation/MergeOccurences/@EntryValue">True</s:Boolean>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateBlankLinesAroundFieldToBlankLinesAroundProperty/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ESettingsUpgrade_002EMigrateThisQualifierSettings/@EntryIndexedValue">True</s:Boolean>
```
Comment (amaitland) [inline CefSharp3.sln.DotSettings]: Can you revert this change please? I'm guessing this was just an oversight.

**Comment by amaitland** [comment]
@daveaton I see you made the changes :+1: I'll take this for a spin ASAP :smile: 

(When you push more commits github doesn't notify me, so as a future reference if you could post a comment when changes are complete that would be much appreciated).

**Comment by daveaton** [comment]
@amaitland Sorry about that I did not want to seem pushy.  I normally check stats about 1-3 times a day.  I figured you were hard at work on v39 :)  Now that I know I will start leaving comments.

**Comment by amaitland** [comment]
@daveaton Just a quick, `changes made, let me know if there's any further feedback` would be perfect :smile:

**Comment by daveaton** [comment]
@amaitland Will do :)

**Comment by amaitland** [comment]
@daveaton I merged this, made a few additional changes. You can review them here https://github.com/cefsharp/CefSharp/commit/fce5f06362fb85fe63792e376e2bfd747d46c689

Basically I removed a few unnessicary params (The Js callback in `Cef` is generic, so only the first param was required).

I've tested with some basic scenarios, everything appears to be working. If you've got any comments please let me know :smile:
